bids/layout/layout.py

```diff
@@ -598,62 +856,130 @@ def get_tr(self, derivatives=False, **selectors):
-            types (str, list): Types of variables to retrieve. All valid values
-            reflect the filename stipulated in the BIDS spec for each kind of
-            variable. Valid values include: 'events', 'physio', 'stim',
-            'scans', 'participants', 'sessions', and 'regressors'.
-            variables (list): Optional list of variables names to return. If
-                None, all available variables are returned.
-            merge (bool): If True, variables are merged across all observations
-                of the current level. E.g., if level='subject', variables from
-                all subjects will be merged into a single collection. If False,
-                each observation is handled separately, and the result is
-                returned as a list.
-            sampling_rate (int, str): If level='run', the sampling rate to
-                pass onto the returned BIDSRunVariableCollection.
-            skip_empty (bool): Whether or not to skip empty Variables (i.e.,
-                where there are no rows/records in a file after applying any
-                filtering operations like dropping NaNs).
-            kwargs: Optional additional arguments to pass onto load_variables.
-        """
-        from bids.variables import load_variables
-        index = load_variables(self, types=types, levels=level,
-                               skip_empty=skip_empty, **kwargs)
-        return index.get_collections(level, variables, merge,
-                                     sampling_rate=sampling_rate)
+            source (str, File, dict): The source data to use to construct the
+                new file path. Must be one of:
+                - A File object
+                - A string giving the path of a File contained within the
+                  current Layout.
+                - A dict of entities, with entity names in keys and values in
+                  values
+            path_patterns (list): Optional path patterns to use to construct
+                the new file path. If None, the Layout-defined patterns will
+                be used.
+            strict (bool): If True, all entities must be matched inside a
+                pattern in order to be a valid match. If False, extra entities
+                will be ignored so long as all mandatory entities are found.
+            scope (str, list): The scope of the search space. Indicates which
+                BIDSLayouts' path patterns to use. See BIDSLayout docstring
+                for valid values. By default, uses all available layouts. If
+                two or more values are provided, the order determines the
+                precedence of path patterns (i.e., earlier layouts will have
+                higher precedence).
+        '''
 
-    def _make_file_object(self, root, f):
-        # Override grabbit's File with a BIDSFile.
-        return BIDSFile(os.path.join(root, f), self)
+        if isinstance(source, six.string_types):
```

Comment:
Do we want to consider `Path`-like objects?

Reply:
What's diagnostic of a `Path`-like object? Are you okay with checking for an attribute that only a `PurePath` would have, or is it cool if I just check for `__str__`? (The risk in the latter is that in principle all kinds of objects could have sensible string representations that aren't paths. But if you're passing in random objects to `build_path`, I'm happy to file that under user error.)

Reply:
I think in the other place we're doing it we just `try: source = str(source) except: pass`

Reply:
#307, for quick reference.

Reply:
I saw that, but in that case we just fail, whereas here the user could still be passing in a `BIDSFile`, which is valid but doesn't have a `__str__` method. So I'll check for `__str__`, which is equivalent.

Reply:
Actually, I take it back, that works fine.

Reply:
isinstance?

Reply:
This is getting more involved than it needs to be. I'm sticking with the `hasattr(source, "__str__")` check because if we use try/except we risk swallowing a different error coming from an `os.path.join` call, and nesting try/excepts is overkill.

Reply:
Argh, now that's failing because of course some types have string representations but aren't hashable and can't be checked against `self.files` on the next line. Jesus, this is not what I expected to take me down a rabbit hole.

Reply:
LOL, let's leave it as a `help-wanted` issue.

Reply:
I'm just going to check for `hasattr(source, "is_file")`. That seems like a pretty good tell if something is a `Path` object. If anyone objects, submit a PR. :p

Reply:
Pretty much everybody using `Path`s is using Python 3, so I don't think we need to worry overly much about the variety of things that are `Path`-like.

Reply:
This also means that that earlier `self.root = str(self.root)` call is broken, because if you pass a dictionary as `root`, it'll convert it to a string, but won't fail until later. The rabbit hole goes deeper!

Reply:
@yarikoptic, please add this to the list you're compiling of "annoying little things that would be trivial for us to deal with if @yarikoptic allowed us to drop Python 2 support." ;)

Reply:
Maybe we can abstract this out like we do with `listify` and do:

```Python
def pathify(maybe_path):
    if isinstance(maybe_path, six.string_types):
        return maybe_path
    if hasattr(maybe_path, 'is_file') or hasattr(maybe_path, '__fspath__'):
        return str(maybe_path)
    # any others?
    ...
    # Return to let caller decide what to do with non-pathy objects
    return maybe_path
```